src/lib/reasoners/arrays_rel.ml

```diff
@@ -415,17 +419,40 @@ let new_equalities env eqs la class_of =
   implied_consequences env eqs la
 
 (* choisir une egalite sur laquelle on fait un case-split *)
-let two = Numbers.Q.from_int 2
-
-let case_split env _uf ~for_model:_ =
+let two = Numbers.Q.from_int 3
+
+let split_arrays ~for_model env =
+  if for_model then
+    match SX.choose env.arrays with
+    | exception Not_found -> None
+    | a1 ->
+      match SX.find_first (fun a -> not @@ X.equal a a1) env.arrays with
```

Comment:
This needs to take into consideration the types of arrays. We must not create an equality literal between arrays of different types. Probably the `arrays` table needs to be segregated by both index and value types.

Reply:
Also, the current implementation loops trying to assert the same disequality. See for instance:

```smt-lib
(set-logic ALL)
(set-option :produce-models true)
(declare-const a (Array Int Int))
(declare-const b (Array Int Int))
(declare-const c (Array Int Int))
(declare-const i Int)
(assert (= (select a i) (select b i)))
(assert (= (select a i) (select c i)))
(check-sat)
(get-model)
```

With `-d split` we can see that we loop repeatedly selecting `b <> a` as a split during model generation. `Uf.already_distinct` would need to be checked.